core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,322 @@
+ */
+public class UniformPartitionRebalancer
+{
+    private static final Logger log = Logger.get(UniformPartitionRebalancer.class);
+    private static final long MIN_REBALANCE_PHYSICAL_WRITTEN_BYTES = DataSize.of(100, MEGABYTE).toBytes();
+
+    // If the percentage difference between the two writers with maximum and minimum physical written bytes
+    // since last rebalance is above 0.7 (or 70%), then we consider them skewed.
+    private static final double SKEWNESS_THRESHOLD = 0.7;
+
+    private final List<Supplier<Long>> writerPhysicalWrittenBytesSuppliers;
+    private final long writerMinSize;
+    private final int numberOfWriters;
+
+    @GuardedBy("this")
+    private final PartitionWriterInfo[] partitionWriterInfos;
+
+    @GuardedBy("this")
+    private final List<IndexedPriorityQueue<Integer>> writerPartitionIdWithRowCount = new ArrayList<>();
+
+    @GuardedBy("this")
+    private final long[] lastRefreshWriterPhysicalWrittenBytes;
+
+    public UniformPartitionRebalancer(
+            List<Supplier<Long>> writerPhysicalWrittenBytesSuppliers,
+            int partitionCount,
+            int numberOfWriters,
+            long writerMinSize)
+    {
+        this.writerPhysicalWrittenBytesSuppliers = requireNonNull(writerPhysicalWrittenBytesSuppliers, "writerPhysicalWrittenBytesSuppliers is null");
+        this.writerMinSize = writerMinSize;
+        this.numberOfWriters = numberOfWriters;
+
+        lastRefreshWriterPhysicalWrittenBytes = new long[numberOfWriters];
+        for (int i = 0; i < numberOfWriters; i++) {
+            lastRefreshWriterPhysicalWrittenBytes[i] = 0;
+            writerPartitionIdWithRowCount.add(new IndexedPriorityQueue<>());
+        }
+
+        partitionWriterInfos = new PartitionWriterInfo[partitionCount];
+        for (int i = 0; i < partitionWriterInfos.length; i++) {
+            partitionWriterInfos[i] = new PartitionWriterInfo(i % numberOfWriters);
+        }
+    }
+
+    public synchronized void rebalancePartitions()
+    {
+        long[] writerWrittenBytes = new long[numberOfWriters];
+        long[] writerWrittenBytesSinceLastRebalance = new long[numberOfWriters];
+        long totalWrittenBytesSinceLastRebalance = 0;
+        for (int writerId = 0; writerId < numberOfWriters; writerId++) {
+            writerWrittenBytes[writerId] = writerPhysicalWrittenBytesSuppliers.get(writerId).get();
+            long writtenBytesSinceLastRebalance = writerWrittenBytes[writerId] - lastRefreshWriterPhysicalWrittenBytes[writerId];
+
+            writerWrittenBytesSinceLastRebalance[writerId] = writtenBytesSinceLastRebalance;
+            totalWrittenBytesSinceLastRebalance += writtenBytesSinceLastRebalance;
+        }
+
+        // Rebalance only when total bytes written since last rebalance is greater than MIN_PHYSICAL_WRITTEN_REBALANCE_BYTES (100 MB)
+        if (totalWrittenBytesSinceLastRebalance <= MIN_REBALANCE_PHYSICAL_WRITTEN_BYTES) {
```

Comment:
Hmm, actually there is a shortcut. I wonder if it makes sense to make it more lightweight (do it before any allocations are done, also maybe even outside the lock)